### PR TITLE
[oc] Auto create TLS cert in collector deployment

### PR DESCRIPTION
**Pull request by annanay25**
For OpenShift platform -
- Annotate the collector service with tag `service.beta.openshift.io/serving-cert-secret-name`
- Mount the auto-created secret (with the crt and key)on the pod and use it in the cli for the collector grpc server

Resolves: #599 

Signed-off-by: Annanay <annanayagarwal@gmail.com>

### CODE REVIEW

**Comment by jpkrohling** [comment]
cc @kevinearls, @jkandasa could one of you please give this a try?

File: pkg/config/tls/tls.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	cm := &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-collector-tls-config", t.jaeger.Name),
+			Namespace: t.jaeger.Namespace,
+			Labels: map[string]string{
+				"app":                          "jaeger",
```
[the remark this answers is not in the document]
Comment (annanay25) [reply pkg/config/tls/tls.go]: Done.

**Comment by annanay25** [comment]
Thanks for the review @jpkrohling. I've addressed comments :)

**Comment by jpkrohling** [comment]
Looks good! The mentioned PR has been merged, so, this can be rebased against master again.

@kevinearls, @jkandasa, could one of you please test this?

File: pkg/config/tls/tls.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	if viper.GetString("platform") != v1.FlagPlatformOpenShift {
+		return
+	}
+
+	volume := corev1.Volume{
+		Name: configurationVolumeName(jaeger),
+		VolumeSource: corev1.VolumeSource{
+			ConfigMap: &corev1.ConfigMapVolumeSource{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: fmt.Sprintf("%s-tls-configuration", jaeger.Name),
```
Comment (jpkrohling) [inline pkg/config/tls/tls.go]: This should probably be `collector-tls-config` instead of `tls-configuration`:

```
$ kubectl get configmaps
NAME                                 DATA   AGE
simple-prod-collector-tls-config     1      5m17s
simple-prod-sampling-configuration   1      5m17s
simple-prod-ui-configuration         1      5m17s
```

File: pkg/config/tls/tls.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	}
+	volumeMount := corev1.VolumeMount{
+		Name:      configurationVolumeName(jaeger),
+		MountPath: "/etc/config",
+		ReadOnly:  true,
+	}
+	commonSpec.Volumes = append(commonSpec.Volumes, volume)
+	commonSpec.VolumeMounts = append(commonSpec.VolumeMounts, volumeMount)
+	*options = append(*options, "--collector.grpc.tls=true")
+	*options = append(*options, "--collector.grpc.tls.cert=/etc/config/service-ca.crt")
```
Comment (jpkrohling) [inline pkg/config/tls/tls.go]: Interesting, I get this: 

```
{"level":"fatal","ts":1582211545.4032314,"caller":"collector/main.go:135","msg":"Could not start gRPC collector","error":"for client auth via TLS, either both client certificate and key must be supplied, or neither","stacktrace":"main.main.func1\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/collector/main.go:135\ngithub.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra.(*Command).execute\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra/command.go:698\ngithub.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra.(*Command).ExecuteC\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra/command.go:783\ngithub.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra.(*Command).Execute\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra/command.go:736\nmain.main\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/collector/main.go:183\nruntime.main\n\t/home/travis/.gimme/versions/go1.13.5.linux.amd64/src/runtime/proc.go:203"}
```

**Comment by jpkrohling** [comment]
@annanay25 do you need a cluster to test this?

File: pkg/config/tls/tls.go (hunk 2 above)
Comment (annanay25) [reply pkg/config/tls/tls.go]: Will do.

File: pkg/config/tls/tls.go (hunk 3 above)
Comment (annanay25) [reply pkg/config/tls/tls.go]: Ha, I'm actually not sure how to fix this? Can I generate and supply a specific tls key to a container through openshift?

Comment (annanay25) [reply pkg/config/tls/tls.go]: Update: According to [this link](https://docs.openshift.com/container-platform/4.1/authentication/certificates/service-serving-certificate.html#add-service-certificate_service-serving-certificate) on the original issue, it seems like the secret will contain both the cert and key.

**Comment by annanay25** [comment]
> do you need a cluster to test this?

Yes, that would be great @jpkrohling :) . I need to confirm if `service.beta.openshift.io/inject-cabundle=true` injects both the cert and the key (I think it should, since it says ca-bundle).

**Comment by jpkrohling** [comment]
Just sent you the details to your OpenShift cluster in private, via Gitter.

**Comment by annanay25** [comment]
There seems to be one warning - 

```
Annanays-Mac:jaeger-operator annanay$ kubectl logs -f pod/simple-prod-collector-649fddfc7c-xxd65
2020/02/25 14:21:03 maxprocs: Leaving GOMAXPROCS=2: CPU quota undefined
{"level":"info","ts":1582640463.727751,"caller":"flags/service.go:115","msg":"Mounting metrics handler on admin server","route":"/metrics"}
{"level":"info","ts":1582640463.728069,"caller":"flags/admin.go:108","msg":"Mounting health check on admin server","route":"/"}
{"level":"info","ts":1582640463.7281487,"caller":"flags/admin.go:114","msg":"Starting admin HTTP server","http-port":14269}
{"level":"info","ts":1582640463.7281687,"caller":"flags/admin.go:100","msg":"Admin server started","http-port":14269,"health-status":"unavailable"}
{"level":"info","ts":1582640463.783358,"caller":"config/config.go:172","msg":"Elasticsearch detected","version":5}
{"level":"info","ts":1582640463.8457026,"caller":"collector/main.go:128","msg":"Starting jaeger-collector TChannel server","port":14267}
{"level":"warn","ts":1582640463.8457642,"caller":"collector/main.go:129","msg":"TChannel has been deprecated and will be removed in a future release"}
{"level":"info","ts":1582640463.9107623,"caller":"grpcserver/grpc_server.go:64","msg":"Starting jaeger-collector gRPC server","grpc-port":"14250"}
{"level":"info","ts":1582640463.9108763,"caller":"collector/main.go:148","msg":"Starting jaeger-collector HTTP server","http-port":14268}
{"level":"info","ts":1582640463.9109273,"caller":"healthcheck/handler.go:128","msg":"Health Check state change","status":"ready"}
{"level":"info","ts":1582640463.936088,"caller":"collector/main.go:240","msg":"Listening for Zipkin HTTP traffic","zipkin.http-port":9411}
WARNING: 2020/02/25 14:21:32 grpc: Server.Serve failed to complete security handshake from "10.129.2.19:60554": tls: first record does not look like a TLS handshake
WARNING: 2020/02/25 14:21:33 grpc: Server.Serve failed to complete security handshake from "10.129.2.19:60560": tls: first record does not look like a TLS handshake
```

File: pkg/deployment/agent.go [hunk 4]
```diff
@@ -43,6 +44,11 @@ func (a *Agent) Get() *appsv1.DaemonSet {
 		if len(util.FindItem("--reporter.grpc.host-port=", args)) == 0 {
 			args = append(args, fmt.Sprintf("--reporter.grpc.host-port=dns:///%s.%s:14250", service.GetNameForHeadlessCollectorService(a.jaeger), a.jaeger.Namespace))
```
Comment (jpkrohling) [inline pkg/deployment/agent.go]: There's something odd: this would generate a string like: `- --reporter.grpc.host-port=dns:///simple-prod-collector-headless.default:14250` . Yet, the cert is valid only for the service `simple-prod-collector`:

```console
$ kubectl get secrets jaeger-collector-tls -o=go-template='{{index .data "tls.crt"}}' | base64 -d | openssl x509 -text -noout
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 7784809259584540342 (0x6c0932c6a1b54eb6)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = openshift-service-serving-signer@1582647274
        Validity
            Not Before: Feb 25 16:37:59 2020 GMT
            Not After : Feb 24 16:38:00 2022 GMT
        Subject: CN = simple-prod-collector.default.svc
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
                Modulus:
                    00:d9:1c:19:e9:68:d4:17:fb:50:84:ba:32:7c:27:
                    ed:eb:f7:28:9e:76:e2:e9:5b:5f:73:12:31:24:49:
                    6f:1f:8d:1e:07:26:58:64:39:4a:48:1e:7d:f6:33:
                    02:38:17:6b:72:40:6c:4d:93:e6:d8:89:d8:07:f1:
                    40:46:fe:da:5e:dc:8c:de:21:1d:77:64:49:4f:8d:
                    3b:5b:56:70:da:a0:32:0a:48:75:d3:df:7f:77:72:
                    c4:42:09:28:db:70:93:dd:d9:cb:b6:d1:45:b5:7b:
                    38:39:a5:09:57:c5:cd:80:84:64:5f:66:2c:45:e1:
                    cc:53:f0:90:df:d9:09:a9:c8:b2:0d:19:c0:25:92:
                    f0:8a:35:1c:ab:3b:2e:21:06:c2:0e:57:38:90:95:
                    df:17:00:66:50:bb:80:b0:78:fe:12:35:bf:4b:26:
                    7e:a2:2c:38:d6:83:3f:87:42:c8:c7:12:bc:4a:fc:
                    13:01:ae:13:ba:94:9b:e1:de:37:e9:a6:d8:79:02:
                    57:21:8c:49:6e:4b:05:13:a7:b2:fd:17:08:3b:ce:
                    f4:b9:a1:35:39:df:6e:58:3b:57:47:b6:89:c3:95:
                    bb:65:0c:fd:e6:68:89:d9:63:09:82:4d:f1:00:97:
                    72:26:9e:34:60:5e:24:30:83:7b:b3:40:19:ce:8a:
                    58:9f
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Subject Key Identifier: 
                71:90:9D:90:2F:DB:63:5E:E6:53:24:50:FB:54:23:1E:DD:26:3A:5C
            X509v3 Authority Key Identifier: 
                keyid:F8:A9:03:4A:FD:7B:74:9F:D3:8E:F3:ED:B9:B7:D3:DC:E7:F5:85:5D

            X509v3 Subject Alternative Name: 
                DNS:simple-prod-collector.default.svc, DNS:simple-prod-collector.default.svc.cluster.local
            1.3.6.1.4.1.2312.17.100.2.1: 
                .$47b497a9-bbad-46a4-b5d9-6330f2c186de
    Signature Algorithm: sha256WithRSAEncryption
         7e:9e:ac:f4:63:99:0a:d3:6b:07:3c:e2:3e:e6:86:0a:b5:0a:
         b3:04:84:af:d1:fb:04:2d:1a:97:f5:f2:23:03:fe:e2:f3:35:
         17:6f:01:53:75:89:ac:a5:22:80:9d:63:e1:c2:38:aa:b9:64:
         f8:9c:ec:e9:cd:17:77:4e:b0:d5:06:99:3a:cc:c4:43:d8:5a:
         99:9f:12:fd:24:6f:e3:79:89:89:7c:1e:f1:8b:81:5e:c7:fc:
         15:25:f1:8f:12:55:b6:3c:03:23:24:3a:83:dc:eb:f1:52:34:
         d6:d0:87:1c:ce:68:0e:e5:51:9a:5b:f6:3c:5e:31:f2:9c:a0:
         51:6c:4d:8f:79:2b:c4:a0:14:f5:7d:1d:6c:c6:66:d5:b6:27:
         df:78:8e:0c:43:68:a9:86:55:81:2e:e1:4f:03:25:c8:0c:03:
         f2:8f:f0:2f:c6:f9:b2:a0:7d:d6:fb:23:8f:d9:cd:7b:a6:d2:
         cb:17:65:66:ea:66:62:bd:00:10:cb:93:60:09:fe:d9:da:d8:
         bb:3a:a4:ff:85:fa:4b:4e:28:e0:d1:f0:45:8a:aa:45:0a:b9:
         35:d0:4d:a0:75:ff:17:01:9d:86:5b:32:4f:50:88:ca:bb:83:
         8a:41:ab:34:40:1b:5d:33:43:5a:17:8a:16:2b:60:4a:2d:ab:
         5e:4a:4e:56
```

This would explain why I don't see traces from the Jaeger UI being shown there :-)

Comment (jpkrohling) [inline pkg/deployment/agent.go]: There's something odd: this would generate a string like: `- --reporter.grpc.host-port=dns:///simple-prod-collector-headless.default:14250` . Yet, the cert is valid only for the service `simple-prod-collector`:

```console
$ kubectl get secrets jaeger-collector-tls -o=go-template='{{index .data "tls.crt"}}' | base64 -d | openssl x509 -text -noout
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 7784809259584540342 (0x6c0932c6a1b54eb6)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = openshift-service-serving-signer@1582647274
        Validity
            Not Before: Feb 25 16:37:59 2020 GMT
            Not After : Feb 24 16:38:00 2022 GMT
        Subject: CN = simple-prod-collector.default.svc
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
                Modulus:
                    00:d9:1c:19:e9:68:d4:17:fb:50:84:ba:32:7c:27:
                    ed:eb:f7:28:9e:76:e2:e9:5b:5f:73:12:31:24:49:
                    6f:1f:8d:1e:07:26:58:64:39:4a:48:1e:7d:f6:33:
                    02:38:17:6b:72:40:6c:4d:93:e6:d8:89:d8:07:f1:
                    40:46:fe:da:5e:dc:8c:de:21:1d:77:64:49:4f:8d:
                    3b:5b:56:70:da:a0:32:0a:48:75:d3:df:7f:77:72:
                    c4:42:09:28:db:70:93:dd:d9:cb:b6:d1:45:b5:7b:
                    38:39:a5:09:57:c5:cd:80:84:64:5f:66:2c:45:e1:
                    cc:53:f0:90:df:d9:09:a9:c8:b2:0d:19:c0:25:92:
                    f0:8a:35:1c:ab:3b:2e:21:06:c2:0e:57:38:90:95:
                    df:17:00:66:50:bb:80:b0:78:fe:12:35:bf:4b:26:
                    7e:a2:2c:38:d6:83:3f:87:42:c8:c7:12:bc:4a:fc:
                    13:01:ae:13:ba:94:9b:e1:de:37:e9:a6:d8:79:02:
                    57:21:8c:49:6e:4b:05:13:a7:b2:fd:17:08:3b:ce:
                    f4:b9:a1:35:39:df:6e:58:3b:57:47:b6:89:c3:95:
                    bb:65:0c:fd:e6:68:89:d9:63:09:82:4d:f1:00:97:
                    72:26:9e:34:60:5e:24:30:83:7b:b3:40:19:ce:8a:
                    58:9f
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Subject Key Identifier: 
                71:90:9D:90:2F:DB:63:5E:E6:53:24:50:FB:54:23:1E:DD:26:3A:5C
            X509v3 Authority Key Identifier: 
                keyid:F8:A9:03:4A:FD:7B:74:9F:D3:8E:F3:ED:B9:B7:D3:DC:E7:F5:85:5D

            X509v3 Subject Alternative Name: 
                DNS:simple-prod-collector.default.svc, DNS:simple-prod-collector.default.svc.cluster.local
            1.3.6.1.4.1.2312.17.100.2.1: 
                .$47b497a9-bbad-46a4-b5d9-6330f2c186de
    Signature Algorithm: sha256WithRSAEncryption
         7e:9e:ac:f4:63:99:0a:d3:6b:07:3c:e2:3e:e6:86:0a:b5:0a:
         b3:04:84:af:d1:fb:04:2d:1a:97:f5:f2:23:03:fe:e2:f3:35:
         17:6f:01:53:75:89:ac:a5:22:80:9d:63:e1:c2:38:aa:b9:64:
         f8:9c:ec:e9:cd:17:77:4e:b0:d5:06:99:3a:cc:c4:43:d8:5a:
         99:9f:12:fd:24:6f:e3:79:89:89:7c:1e:f1:8b:81:5e:c7:fc:
         15:25:f1:8f:12:55:b6:3c:03:23:24:3a:83:dc:eb:f1:52:34:
         d6:d0:87:1c:ce:68:0e:e5:51:9a:5b:f6:3c:5e:31:f2:9c:a0:
         51:6c:4d:8f:79:2b:c4:a0:14:f5:7d:1d:6c:c6:66:d5:b6:27:
         df:78:8e:0c:43:68:a9:86:55:81:2e:e1:4f:03:25:c8:0c:03:
         f2:8f:f0:2f:c6:f9:b2:a0:7d:d6:fb:23:8f:d9:cd:7b:a6:d2:
         cb:17:65:66:ea:66:62:bd:00:10:cb:93:60:09:fe:d9:da:d8:
         bb:3a:a4:ff:85:fa:4b:4e:28:e0:d1:f0:45:8a:aa:45:0a:b9:
         35:d0:4d:a0:75:ff:17:01:9d:86:5b:32:4f:50:88:ca:bb:83:
         8a:41:ab:34:40:1b:5d:33:43:5a:17:8a:16:2b:60:4a:2d:ab:
         5e:4a:4e:56
```

This would explain why I don't see traces from the Jaeger UI being shown there :-)

Comment (annanay25) [reply pkg/deployment/agent.go]: Changed to use the headless service.

Comment (jpkrohling) [reply pkg/deployment/agent.go]: Thanks, and sorry for the spam. GitHub was having some problems yesterday and it apparently added one comment per attempt.

**Comment by jpkrohling** [comment]
Looks like it's still not quite working yet. This is what I see in the Agent logs for the Query:

```
{"level":"info","ts":1582707258.2944024,"caller":"base/balancer.go:83","msg":"base.baseBalancer: got new ClientConn state: {{[{10.129.2.18:14250 0  <nil>}] <nil>} <nil>}","system":"grpc","grpc_log":true}
{"level":"info","ts":1582707260.305737,"caller":"base/balancer.go:140","msg":"base.baseBalancer: handle SubConn state change: 0xc0002256b0, CONNECTING","system":"grpc","grpc_log":true}
{"level":"info","ts":1582707260.3057969,"caller":"roundrobin/roundrobin.go:50","msg":"roundrobinPicker: newPicker called with readySCs: map[]","system":"grpc","grpc_log":true}
{"level":"info","ts":1582707260.3066912,"caller":"base/balancer.go:140","msg":"base.baseBalancer: handle SubConn state change: 0xc0002256b0, TRANSIENT_FAILURE","system":"grpc","grpc_log":true}
{"level":"info","ts":1582707260.3067415,"caller":"transport/log.go:30","msg":"transport: loopyWriter.run returning. connection error: desc = \"transport is closing\"","system":"grpc","grpc_log":true}
```

And this is from the collector logs:
```
$ kubectl logs simple-prod-collector-5d556d5db-dxmpr
2020/02/26 08:50:59 maxprocs: Leaving GOMAXPROCS=2: CPU quota undefined
{"level":"info","ts":1582707059.8038766,"caller":"flags/service.go:115","msg":"Mounting metrics handler on admin server","route":"/metrics"}
{"level":"info","ts":1582707059.804306,"caller":"flags/admin.go:108","msg":"Mounting health check on admin server","route":"/"}
{"level":"info","ts":1582707059.8043773,"caller":"flags/admin.go:114","msg":"Starting admin HTTP server","http-port":14269}
{"level":"info","ts":1582707059.8044014,"caller":"flags/admin.go:100","msg":"Admin server started","http-port":14269,"health-status":"unavailable"}
{"level":"info","ts":1582707059.8288152,"caller":"config/config.go:172","msg":"Elasticsearch detected","version":6}
{"level":"info","ts":1582707059.917191,"caller":"collector/main.go:128","msg":"Starting jaeger-collector TChannel server","port":14267}
{"level":"warn","ts":1582707059.917238,"caller":"collector/main.go:129","msg":"TChannel has been deprecated and will be removed in a future release"}
{"level":"info","ts":1582707059.9429655,"caller":"grpcserver/grpc_server.go:64","msg":"Starting jaeger-collector gRPC server","grpc-port":"14250"}
{"level":"info","ts":1582707059.9430513,"caller":"collector/main.go:148","msg":"Starting jaeger-collector HTTP server","http-port":14268}
{"level":"info","ts":1582707059.9430842,"caller":"healthcheck/handler.go:128","msg":"Health Check state change","status":"ready"}
{"level":"info","ts":1582707059.9643774,"caller":"collector/main.go:240","msg":"Listening for Zipkin HTTP traffic","zipkin.http-port":9411}
WARNING: 2020/02/26 08:51:18 grpc: Server.Serve failed to complete security handshake from "10.129.2.19:38274": tls: first record does not look like a TLS handshake
WARNING: 2020/02/26 08:51:19 grpc: Server.Serve failed to complete security handshake from "10.129.2.19:38278": tls: first record does not look like a TLS handshake
WARNING: 2020/02/26 08:51:20 grpc: Server.Serve failed to complete security handshake from "10.129.2.19:38294": tls: first record does not look like a TLS handshake
```

I just got all the certs locally, and I was able to get a working setup. Looks like you are just missing the Agent parameter specifying the service-ca bundle, which makes the agent trust the certs generated by OpenShift's CA. Here's how to test it locally:

```console
$ kubectl get secrets simple-prod-collector-headless-tls -o=go-template='{{index .data "tls.crt"}}' | base64 -d > /tmp/cert.crt

$ kubectl get secrets simple-prod-collector-headless-tls -o=go-template='{{index .data "tls.key"}}' | base64 -d > /tmp/cert.key

$ kubectl apply -f deploy/examples/business-application-injected-sidecar.yaml 

$ kubectl exec -c myapp myapp-7cb8b69d-nvvkv -- cat /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt > /tmp/service-ca.crt

$ SPAN_STORAGE_TYPE=memory go run -tags ui ./cmd/collector --collector.grpc.tls.enabled=true --collector.grpc.tls.cert=/tmp/cert.crt --collector.grpc.tls.key=/tmp/cert.key 

$ sudo vi /etc/hosts # add a line with the cert's hostname: 127.0.0.1 simple-prod-collector-headless.default.svc

$ SPAN_STORAGE_TYPE=memory go run -tags ui ./cmd/agent --reporter.grpc.host-port=simple-prod-collector-headless.default.svc:14250 --reporter.grpc.tls.enabled=true --reporter.grpc.tls.ca=/tmp/service-ca.crt
```

In the last command, you should see this in the logs:

```
{"level":"info","ts":1582708364.6002507,"caller":"base/balancer.go:181","msg":"base.baseBalancer: handle SubConn state change: 0xc0002882b0, READY","system":"grpc","grpc_log":true}
{"level":"info","ts":1582708364.6003134,"caller":"roundrobin/roundrobin.go:48","msg":"roundrobinPicker: newPicker called with info: {map[0xc0002882b0:{{simple-prod-collector-headless.default.svc:14250  <nil> 0 <nil>}}]}","system":"grpc","grpc_log":true}
```

TL;DR: it should be sufficient to add the following arg to the agent: `--reporter.grpc.tls.ca=/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt`

**Comment by annanay25** [comment]
Thanks @jpkrohling for the analysis. I've updated the PR.

**Comment by annanay25** [comment]
Just added https://github.com/jaegertracing/jaeger-operator/pull/914/files#diff-acd8a9fe2d2997867f151eb11780379bR138-R146 in `sidecar.go` which was missing previously

File: pkg/deployment/agent.go [hunk 5]
```diff
@@ -45,6 +46,15 @@ func (a *Agent) Get() *appsv1.DaemonSet {
 		}
 	}
 
+	// Enable tls by default for openshift platform
+	if viper.GetString("platform") == v1.FlagPlatformOpenShift {
+		if len(util.FindItem("--reporter.type=grpc", args)) > 0 && len(util.FindItem("--reporter.grpc.tls=true", args)) == 0 {
+			args = append(args, "--reporter.grpc.tls=true")
+			args = append(args, "--reporter.grpc.tls.ca=/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt")
+			args = append(args, fmt.Sprintf("--reporter.grpc.tls.server-name=%s", service.GetNameForHeadlessCollectorService(a.jaeger)))
```
Comment (jpkrohling) [inline pkg/deployment/agent.go]: Looks like you need to qualify the DNS name:

```
{"level":"info","ts":1583247523.9721072,"caller":"grpc/clientconn.go:1303","msg":"grpc: addrConn.createTransport failed to connect to {10.128.0.96:14250  <nil> 0 <nil>}. Err :connection error: desc = \"transport: authentication handshake failed: x509: certificate is valid for simple-prod-collector-headless.default.svc, simple-prod-collector-headless.default.svc.cluster.local, not simple-prod-collector-headless\". Reconnecting...","system":"grpc","grpc_log":true}
```

```
 			args = append(args, fmt.Sprintf("--reporter.grpc.tls.server-name=%s", service.GetNameForHeadlessCollectorService(a.jaeger)))
```

Becomes:
```
 			args = append(args, fmt.Sprintf("--reporter.grpc.tls.server-name=%s.%s.svc.cluster.local", service.GetNameForHeadlessCollectorService(a.jaeger), a.jaeger.Namespace))
```

File: pkg/inject/sidecar.go [hunk 6]
```diff
@@ -134,6 +135,15 @@ func container(jaeger *v1.Jaeger, dep *appsv1.Deployment) corev1.Container {
 		}
 	}
 
+	// Enable tls by default for openshift platform
+	if viper.GetString("platform") == v1.FlagPlatformOpenShift {
+		if len(util.FindItem("--reporter.type=grpc", args)) > 0 && len(util.FindItem("--reporter.grpc.tls=true", args)) == 0 {
+			args = append(args, "--reporter.grpc.tls=true")
+			args = append(args, "--reporter.grpc.tls.ca=/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt")
+			args = append(args, fmt.Sprintf("--reporter.grpc.tls.server-name=%s", service.GetNameForHeadlessCollectorService(jaeger)))
```
Comment (jpkrohling) [inline pkg/inject/sidecar.go]: Same comment as for the agent: needs to qualify the name with namespace + `svc.cluster.local`.

**Comment by annanay25** [comment]
Done :)

File: pkg/strategy/production.go [hunk 7]
```diff
@@ -71,7 +75,23 @@ func newProductionStrategy(ctx context.Context, jaeger *v1.Jaeger) S {
...
 		}
+
+		// For openshift platform, we want the clusterIP service to be annotated
+		// with the cert secret-name tag
+		for k := range c.services {
+			span.AddEvent(ctx, "Checking service", core.KeyValue{Key: "Servicename", Value: core.String(c.services[k].Name)})
+			if c.services[k].Name == service.GetNameForHeadlessCollectorService(jaeger) {
+				if c.services[k].Annotations == nil {
+					c.services[k].Annotations = map[string]string{
+						"service.beta.openshift.io/serving-cert-secret-name": fmt.Sprintf("%s-tls", c.services[k].Name),
```
Comment (jpkrohling) [inline pkg/strategy/production.go]: I wonder if it wouldn't be easier to add this directly to the relevant service? We might not even need to check if the platform is openshift, as Kubernetes will just ignore this unknown annotation...

https://github.com/jaegertracing/jaeger-operator/blob/9d6319d0f0e828e2a1781d7d5f6a0369d15439c8/pkg/service/collector.go#L33-L78

Comment (annanay25) [reply pkg/strategy/production.go]: You're right, but its just a small extra check to add it to the service on the openshift platform. It could also confuse k8s users to see that annotation

File: pkg/deployment/collector_test.go [hunk 8]
```diff
@@ -486,3 +486,28 @@ func TestAutoscalersDisabledByExplicitOption(t *testing.T) {
 	// verify
 	assert.Len(t, a, 0)
 }
+
+func TestCollectorOrderOfArgumentsOpenshiftTLS(t *testing.T) {
+	viper.Set("platform", v1.FlagPlatformOpenShift)
```
Comment (jpkrohling) [inline pkg/deployment/collector_test.go]:
```suggestion
 	viper.Set("platform", v1.FlagPlatformOpenShift)
 	defer viper.Reset()
```

File: pkg/deployment/collector_test.go [hunk 9]
```diff
@@ -486,3 +486,28 @@ func TestAutoscalersDisabledByExplicitOption(t *testing.T) {
 	// verify
 	assert.Len(t, a, 0)
 }
+
+func TestCollectorOrderOfArgumentsOpenshiftTLS(t *testing.T) {
```
Comment (jpkrohling) [inline pkg/deployment/collector_test.go]: Is the _order_ of arguments important? Or just that they are there?

File: pkg/deployment/collector_test.go [hunk 10]
```diff
@@ -486,3 +486,28 @@ func TestAutoscalersDisabledByExplicitOption(t *testing.T) {
...
+		"a-option": "a-value",
+		"c-option": "c-value",
+	})
+
+	a := NewCollector(jaeger)
+	dep := a.Get()
+
+	assert.Len(t, dep.Spec.Template.Spec.Containers, 1)
+	assert.Len(t, dep.Spec.Template.Spec.Containers[0].Args, 7)
+	assert.True(t, strings.HasPrefix(dep.Spec.Template.Spec.Containers[0].Args[0], "--a-option"))
```
Comment (jpkrohling) [inline pkg/deployment/collector_test.go]: I think we have other tests with better assertions around this. Take a look at the `util` package, I think we have a utility to get arguments from the options map.

File: pkg/inject/sidecar_test.go [hunk 11]
```diff
@@ -596,3 +596,33 @@ func agentTags(args []string) []string {
 	tagsParam := strings.SplitN(tagsArg, "=", 2)[1]
 	return strings.Split(tagsParam, ",")
 }
+
+func TestSidecarOrderOfArgumentsOpenshiftTLS(t *testing.T) {
+	viper.Set("platform", v1.FlagPlatformOpenShift)
```
Comment (jpkrohling) [inline pkg/inject/sidecar_test.go]:
```suggestion
	viper.Set("platform", v1.FlagPlatformOpenShift)
	defer viper.Reset()
```

File: pkg/deployment/agent_test.go [hunk 12]
```diff
@@ -195,3 +195,33 @@ func TestAgentOverrideReporterType(t *testing.T) {
...
+		"a-option": "a-value",
+		"c-option": "c-value",
+	})
+
+	a := NewAgent(jaeger)
+	dep := a.Get()
+
+	assert.Len(t, dep.Spec.Template.Spec.Containers, 1)
+	assert.Len(t, dep.Spec.Template.Spec.Containers[0].Args, 8)
+	assert.True(t, strings.HasPrefix(dep.Spec.Template.Spec.Containers[0].Args[0], "--a-option"))
```
Comment (jpkrohling) [inline pkg/deployment/agent_test.go]: Same comment as the one for `collector_test`.

File: pkg/config/tls/tls.go [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+		},
+	}
+	volumeMount := corev1.VolumeMount{
+		Name:      configurationVolumeName(jaeger),
+		MountPath: "/etc/config",
+		ReadOnly:  true,
+	}
+	commonSpec.Volumes = append(commonSpec.Volumes, volume)
+	commonSpec.VolumeMounts = append(commonSpec.VolumeMounts, volumeMount)
+	*options = append(*options, "--collector.grpc.tls=true")
```
Comment (jpkrohling) [inline pkg/config/tls/tls.go]: Sorry for not catching this earlier, but this flag has just been deprecated in 1.17:

```console
$ podman run jaegertracing/jaeger-collector:1.17 --help | grep collector.grpc.tls.enabled
      --collector.grpc.tls                              (deprecated) see --collector.grpc.tls.enabled
      --collector.grpc.tls.enabled                      Enable TLS on the server
```

**Comment by annanay25** [comment]
@jpkrohling - Could you please re-run CI on this? I'm not sure if the operator deployment is failing because of this PR

**Comment by jpkrohling** [comment]
Merging, as the tests are passing locally:

```
Running Smoke end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	317.085s
Running Cassandra end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	245.803s
Running Elasticsearch end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	407.445s
Running Self provisioned Elasticsearch end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	0.798s
Running Streaming end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	286.869s
Running Example end-to-end tests part 1...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	123.368s
Running Example end-to-end tests part 2...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	164.915s
```

Note that the last one had to be executed twice, because of #945.

**Comment by annanay25** [comment]
Thanks! 🎉
